---
title: "eDNA SDERP: Simulations"
author: "Ryan Peek"
date: "Updated: `r format(Sys.Date())`"
output: 
  html_document:
    highlight: pygments
    theme: yeti
    toc: yes
    toc_float: yes
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)

suppressPackageStartupMessages({
  library(tidyverse);
  library(sf);
  library(leaflet);
  library(here);
  library(mapview)
})

```

## Overview & Objective

To assess how much eDNA is required to identify the number of individuals in a population, we need to first play around with the parameters that may affect this. There are a number of environmental and physical factors which play a role in how much eDNA is available in a sample, such as individual shedding rates, local DNA decay rates, and movement through water/environment. These are separate of the intent of these simulations.

 >  - We want to know, assuming ideal conditions, how much DNA/sequencing/coverage would it require to actually detect a given individual(s)?

### Simulation General Parameters 

Assuming even sequencing coverage across individuals, what combination of loci/length/coverage/theta gives best chance of detecting **_X_** number of individuals?

**The main parameters of interest:**

 1. **nInd** = number of individuals we want to detect (i.e, 1, 5, 100)
 2. $\theta$  = which relates to loci length, $\mu$, and $N_e$. Can vary between organism/sequences (i.e., 0.1, 1, 10)
 3. **nLoci** = Number of Loci sequenced/analyzed (10, 100, 1000, 10000)
 4. **coverage** = Coverage per locus analyzed
 5. **error** = The error associated with sequencing or miscalls. Helps build in error (singleton haplotypes) into simulations
 
### Specific `bash` Parameters

Need to calculate $\theta$ = 4 * $N_e$ * $\mu$ * $llength$ 

 - $\mu$ = Assuming general mutation rate of 1e^-8^
 - $llength$ = number of basepairs
 - $N_e$ = Coalescent effective size, using conservative est of 25,000
 
**Other Parameters:**

 - $nInd$ = number of individuals
 - $nLoci$ = number of loci analyzed
 - $coverage$ = mean coverage per locus analyzed
 - $ploidy$ = ploidy of individuals
 - $gc$ = gene copies
 - $error$ = percent sequencing error (singleton/miscalled haplos)


### Accounting for Sequencing Error

 - Sequencing error is random, but default keeps novel haplotypes that occur only once
 - Better approach would build in ways to compensate for this, (i.e., filter/ignore singleton haplotypes)

**Our approach:**

 - Only consider 2+ haplotypes
 - 10% error rate, make unique haplotypes
 - run for 100x with 10% error and threshold of 2


### Model Combinations

Vary number of individuals most, because that's we want to know...based on what combinations of things can we detect or not detect diff individuals. Here we use `r 50*2*3*2` possible model combinations.

 - $nInd$ = c(1:50)
 - $nLoci$ = c(100, 1000)
 - $llength$ = c(100, 1000, 10000)
 - $coverage$ = c(10, 100)
 - $error$ = (0.10)
 
## Model Steps

1. Using `ms`, generate segregating site data based on the parameters above. Typical call looks like: `ms $gc $nLoci -t $theta > ms.out` with a 1000 replicates. See [this shell script to do simulations](https://github.com/ryanpeek/edna_gensim/blob/master/code/ms_simulate_haplotypes.sh).
2. Use a `perl` script ([here](https://github.com/ryanpeek/edna_gensim/blob/master/code/ms_sample.pl)) to randomly draw and calculate the number of unique haplotypes from each replicate within each locus in simulated data
3. Using a 10% error rate and a threshold of 2 or greater haplotypes, draw and calculate number of unique haplotypes per `ms` replicate/individual.
4. Run all possible parameter combinations above. See this [R script](https://github.com/ryanpeek/edna_gensim/blob/master/code/01_generate_simulation_calls.R) to generate combinations and the [bash script](https://github.com/ryanpeek/edna_gensim/blob/master/code/ms_runs_sbatch.sh) used to run everything (assuming you have `ms`) on a cluster or computer. To run 600 models (50 individuals) took about 9 hours on the farm cluster at UCD.
4. Using `R`, aggregate this data and visualize as *mean* number of haplotypes vs. Number of Individuals across all possible parameter combos. Script to do this is [here](https://github.com/ryanpeek/edna_gensim/blob/master/code/02_calculate_mean_haplotypes_by_loci.R).

### RESULTS: Sim001-no error

**10X**

```{r 10x}

knitr::include_graphics(paste0(here::here(),"/figs/ms_mean_haplotypes_10x_50nInd.png"))

```

**100X**

```{r 100x}

knitr::include_graphics(paste0(here::here(),"/figs/ms_mean_haplotypes_100x_50nInd.png"))

```

### RESULTS: Sim002-10% error with threshold=2+

 
 
## Forthcoming Options

 - variation in DNA quantity among individuals
 - try other distribution properties (median, mode, min, max)
 - Empirical test, pool DNA & RAD, equal or not?

<!--
## ADDITIONAL RESOURCES 
### From Will

 - for **ms**: use a mutation rate that makes sense given the length of reads that you plan to recover (if the per base is 1x10^-8 and you have loci of 84 bp, you'd use a mutation rate of 8.4x10^-7). The `run_ms.sh` calculates pop-sized scaled mutation rate by default.
 
 - for **coala**: It takes pop-size scaled mutation rate, so just times that by 2*Ne for the theta argument. 

 - [ms pdf and program download here](https://uchicago.app.box.com/s/l3e5uf13tikfjm7e1il1eujitlsjdx13)
 

### Using `skeleSim`**

```{r loadskeleSim, eval=FALSE, echo=FALSE}

library(skeleSim) #https://github.com/christianparobek/skeleSim

# Run the app
skeleSimGUI()
 
```

For a [Sample Size analysis](https://cran.r-project.org/web/packages/skeleSim/vignettes/SampleSize.html), we can play with the following:

 - Number of Populations
 - Migration Rates/Type of Model
 - Seq Lengths
 - Number of Loci
 - Type of Locus


### Using `coala` and `ms`

 - See [here](https://cran.r-project.org/web/packages/coala/vignettes/coala-intro.html)

`coala` is a package written by Staab and Metzler (2016). `ms` is a program written by Hudson (2007). We need to load packages and set `ms` as the simulator. Also see `skelesim` (Parobek et al. 2017).

```{r loadSims, echo=FALSE, eval=FALSE}

library(purrr)
library(tidyverse)
library(coala)
# library(msr) # https://github.com/vsbuffalo/msr, devtools::install_github("vsbuffalo/msr")
library(phyclust) # to use ms sim
activate_ms(priority = 500)

# list simulators
list_simulators()
```

#### First Model

Here we try a simulation of:

 - **sample_size** = 1 population, 5 individuals
 - **loci** = 5
 - **loci_length** = 100

```{r Sim5a5a100, eval=FALSE, echo=FALSE}

# set up the population-size scaled mutation rate (4*N*mu)
# ll <- 84 # loci length
# mu <- 1e-4 # mutation rate
# (ll_mu <- ll*mu) # loci length scaled rate?
# N <- 20 #  pop size
# (pop_size_r <- 4*N*ll_mu)

# 1 Populations, 5 indiv
model <- coal_model(sample_size = c(10), loci_length = 100,
                    loci_number = 100, ploidy = 2) +
  feat_mutation(rate = 1, model = "IFS") + # rate= 4*N0*mu
  #sumstat_nucleotide_div() + 
  sumstat_seg_sites() #+
  #sumstat_sfs() # site freq spectrum, defaults to all pops
  #sumstat_jsfs(name = "jsfs_12", populations = c(1,2)) # joint sfs for 2 pops
  #sumstat_dna() # generates seq data

# check model
check_model(model)
model # view model in ms form

# simulate
ms_sims <- simulate(model, seed = 15, nsim=1)

# setup of list of names
(vecNames <- sprintf("s%03d", 1:10))

# now add to list of sims
ms_sims <- purrr::set_names(ms_sims, nm = vecNames)

# double check
map(ms_sims, ~names(.x))

# get seg sites
get_snps(ms_sims[["s001"]]$seg_sites)



# get just pi
pi_df <- map(ms_sims, ~.x["pi"]) %>% flatten() %>% bind_cols() %>% 
  rename_all(., ~vecNames) %>% 
  pivot_longer(cols = c(1:10), values_to =  "pi")# %>% View()

# ggplot
ggplot() + 
  geom_boxplot(data=pi_df, aes(y=pi, x=name, group=name)) + 
  geom_point(data=pi_df, aes(x = name, y=pi), alpha=.5) +
  labs(x="")

# get seg sites
#ss_df <- 
# this only pulls 3rd in list of seg sites
map(ms_sims, ~pluck(.x[["seg_sites"]][[3]])) %>% simplify()

# test plots
barplot(ms_sims[["s001"]][["sfs"]])
plot(ms_sims[[1]]$sfs)

# plot SFS
(sfs_df <- map(ms_sims, "sfs")  %>% bind_rows())

# single simulation
barplot(sfs_df$sim_001 / sum(sfs_df$sim_001), 
        names.arg = seq_along(sfs_df$sim_001), 
        col = "darkblue")
barplot(sfs_df$sim_004 / sum(sfs_df$sim_004), 
        names.arg = seq_along(sfs_df$sim_004), 
        col = "skyblue", add=TRUE, width = .8)
barplot(sfs_df$sim_003 / sum(sfs_df$sim_003), 
        names.arg = seq_along(sfs_df$sim_003), 
        col = "slateblue", add=TRUE, width = .5)


sfs_out <- rowMeans(sfs_df / sum(sfs_df)) %>% 
  purrr::as_vector()
sfs_df2 <- tibble("sim"=vecNames[-1], sfs_out)
  
ggplot() + geom_col(data=sfs_df2, aes(x=sim, y=sfs_out), fill="cadetblue4")

# plot nucleotide diversity against mutation rates
mean_pi <- sapply(ms_sims, function(x) mean(x$pi))
theta <- sapply(ms_sims, function(x) x$pars[["theta"]])

plot(theta, mean_pi, pch = 19, col = "orange", 
     xlab = "Mutation rate (theta)",
     ylab = "Nucleotide Diversity (mean pi)")
abline(lm(mean_pi ~ theta), col = "gray", lty = 2, lwd=2)


```
 
 
-->
