---
title: "eDNA Simulation Outline"
author: "Ryan Peek"
date: "Updated: `r format(Sys.Date())`"
output: 
  html_document:
    highlight: pygments
    theme: yeti
    toc: yes
    toc_float: yes
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)

suppressPackageStartupMessages({
  library(tidyverse);
  library(sf);
  library(leaflet);
  library(here);
  library(mapview)
})

```

## Overview

To assess how much eDNA is required to identify the number of individuals in a population, we need to first play around with the parameters that may affect this. There are environmental/physical factors which play a role in how much eDNA is available in a sample, such as individual shedding rates, local decay rates, and movement through water/environment. These are separate of the intent of these simulations. We want to know if conditions are ideal, how much DNA would it require to actually detect an individual or a small population? 

To that end there are a few main components or parameters of interest:

 1. The number of individuals we want to detect (i.e, 1, 5, 100)
 2. $\theta$ (mutation rate) which relates to length, and varies between organism/sequences (0.1, 1, 10)
 3. Number of Loci (10, 100, 1000, 10000)?
 
For technical simulations, we'll want to tune with these parameters:
 - Number of loci
 - Length of loci
 - Sequencing depth per loci (more seq, more allelic depth)
 - Number of individuals present
 
*From will*:
 - for **ms**: use a mutation rate that makes sense given the length of reads that you plan to recover (if the per base is 1x10^-8 and you have loci of 84 bp, you'd use a mutation rate of 8.4x10^-7). The `run_ms.sh` calculates pop-sized scaled mutation rate by default.
 
 - for **coala**: It takes pop-size scaled mutation rate, so just times that by 2*Ne for the theta argument. 

 - [ms pdf and program download here](https://uchicago.app.box.com/s/l3e5uf13tikfjm7e1il1eujitlsjdx13)
 
 
 > Assume even coverage across individuals, how many loci/length/coverage gives best chance of detecting **_X_** number of individuals?
 
 Make a plot of # of Alleles (X axis) vs. # of Loci (Y axis).
 
### Using `skeleSim`

```{r loadskeleSim, eval=FALSE}

library(skeleSim) #https://github.com/christianparobek/skeleSim

# Run the app
skeleSimGUI()
 
```

For a [Sample Size analysis](https://cran.r-project.org/web/packages/skeleSim/vignettes/SampleSize.html), we can play with the following:

 - Number of Populations
 - Migration Rates/Type of Model
 - Seq Lengths
 - Number of Loci
 - Type of Locus


### Using `coala` and `ms`

 - See [here](https://cran.r-project.org/web/packages/coala/vignettes/coala-intro.html)

`coala` is a package written by Staab and Metzler (2016). `ms` is a program written by Hudson (2007). We need to load packages and set `ms` as the simulator. Also see `skelesim` (Parobek et al. 2017).

```{r loadSims}

library(purrr)
library(tidyverse)
library(coala)
# library(msr) # https://github.com/vsbuffalo/msr, devtools::install_github("vsbuffalo/msr")
library(phyclust) # to use ms sim
activate_ms(priority = 500)

# list simulators
list_simulators()
```

### First Model

Here we try a simulation of:

 - **sample_size** = 1 population, 5 individuals
 - **loci** = 5
 - **loci_length** = 100

```{r Sim5a5a100}

# set up the population-size scaled mutation rate (4*N*mu)
# ll <- 84 # loci length
# mu <- 1e-4 # mutation rate
# (ll_mu <- ll*mu) # loci length scaled rate?
# N <- 20 #  pop size
# (pop_size_r <- 4*N*ll_mu)

# 1 Populations, 5 indiv
model <- coal_model(sample_size = c(10), loci_length = 100,
                    loci_number = 100, ploidy = 2) +
  feat_mutation(rate = 1, model = "IFS") + # rate= 4*N0*mu
  #sumstat_nucleotide_div() + 
  sumstat_seg_sites() #+
  #sumstat_sfs() # site freq spectrum, defaults to all pops
  #sumstat_jsfs(name = "jsfs_12", populations = c(1,2)) # joint sfs for 2 pops
  #sumstat_dna() # generates seq data

# check model
check_model(model)
model # view model in ms form

# simulate
ms_sims <- simulate(model, seed = 15, nsim=10)

# setup of list of names
(vecNames <- sprintf("s%03d", 1:10))

# now add to list of sims
ms_sims <- purrr::set_names(ms_sims, nm = vecNames)

# double check
map(ms_sims, ~names(.x))

# get seg sites
get_snps(ms_sims[["s001"]]$seg_sites)



# get just pi
pi_df <- map(ms_sims, ~.x["pi"]) %>% flatten() %>% bind_cols() %>% 
  rename_all(., ~vecNames) %>% 
  pivot_longer(cols = c(1:10), values_to =  "pi")# %>% View()

# ggplot
ggplot() + 
  geom_boxplot(data=pi_df, aes(y=pi, x=name, group=name)) + 
  geom_point(data=pi_df, aes(x = name, y=pi), alpha=.5) +
  labs(x="")

# get seg sites
#ss_df <- 
# this only pulls 3rd in list of seg sites
map(ms_sims, ~pluck(.x[["seg_sites"]][[3]])) %>% simplify()

# test plots
barplot(ms_sims[["s001"]][["sfs"]])
plot(ms_sims[[1]]$sfs)

# plot SFS
(sfs_df <- map(ms_sims, "sfs")  %>% bind_rows())

# single simulation
barplot(sfs_df$sim_001 / sum(sfs_df$sim_001), 
        names.arg = seq_along(sfs_df$sim_001), 
        col = "darkblue")
barplot(sfs_df$sim_004 / sum(sfs_df$sim_004), 
        names.arg = seq_along(sfs_df$sim_004), 
        col = "skyblue", add=TRUE, width = .8)
barplot(sfs_df$sim_003 / sum(sfs_df$sim_003), 
        names.arg = seq_along(sfs_df$sim_003), 
        col = "slateblue", add=TRUE, width = .5)


sfs_out <- rowMeans(sfs_df / sum(sfs_df)) %>% 
  purrr::as_vector()
sfs_df2 <- tibble("sim"=vecNames[-1], sfs_out)
  
ggplot() + geom_col(data=sfs_df2, aes(x=sim, y=sfs_out), fill="cadetblue4")

# plot nucleotide diversity against mutation rates
mean_pi <- sapply(ms_sims, function(x) mean(x$pi))
theta <- sapply(ms_sims, function(x) x$pars[["theta"]])

plot(theta, mean_pi, pch = 19, col = "orange", 
     xlab = "Mutation rate (theta)",
     ylab = "Nucleotide Diversity (mean pi)")
abline(lm(mean_pi ~ theta), col = "gray", lty = 2, lwd=2)


```
 
 
# Scaling $\theta$

 - Typically $\theta$ may be around 0.0001 per bp
 - Multiply X bp * 1e-8

```
theta <- 4*N*mu*popsize 
```

Need to scale by bp:      
1 out of 1000 bp differs = theta = 1

## Simulations

### With 10 diploid indiv.

      100bp 1000bp
theta 0.1   1  
n     20    20
nruns 100   100

### With 100 diploid indiv.

      100bp 1000bp
theta 0.1   1  
n     200   200
nruns 100   100

`bash tst_ms.sh 200 200 1 results`

## Next Steps

We want to draw a random locus, and then a random individual:

 - then observe # of unique haplotypes in that individual
 - plot vs. number of total loci possible