---
title: "Meeting Notes"
date: "Updated: `r format(Sys.Date())`"
output: 
  html_document:
    highlight: pygments
    theme: yeti
    toc: yes
    toc_depth: 2
    toc_float: yes
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)

suppressPackageStartupMessages({
  library(tidyverse);
  library(here);
})

```

# Meeting 2020-03-12

## Bucket Experiments

If we are going to run these, supplies we need:

 - no glass (DNA sticks)
 - clean space
 - time to degradation/equilibrium? 
 - Caren recommended 3-4 hours
 - 5 L (treated and ready for fish)
 - most fish DNA 1 - 10 microns
 - filters 0.45 microns (can clog) this would match the underlying distributions of bullfrog experiments
 - Steriltec
 - Start with high density tank first and do 250 mL amounts
 - fold up and stick in a 2ml tube with ethanol
 - or fold in half inward and put in separate coin envelopes with ziploc w/ silica
 - Ship samples to WSU for extraction, WSU sends half of samples.

### Materials

 - https://www.sterlitech.com/autofil-250-ml-0-45-um-funnel-only.html
 - PERMITS: if an amendment, need to run by DOD IACUC, IF NOT, we can use indirect IACUC/permitting
 - SHIPPING samples to WSU
 - **Make sure we are monthly billing (Debbie)**
 
### Replicates

 - 2-3 replicates?
 - 1, 2, 5, 10 (replicated x2),
 - try 40 once and if enough 2x

# Meeting 2020-04-14

### To Do: 

 - discuss updating the sims with diff distributions
    - focus on Poisson, Gaussian, and uniform
 - review paper Caren sent
 - write up manuscript without field experiments for now? 
 - update and submit the regulation paper

### eDNA manuscript

**make a conceptual model:** and use to set up intro and paper. Check Bioscience paper that Sarah wrote for guidance.

 - schematic of different distributions with variables labeled
 
### Distributions

```{r}
# make a pois distrib:
set.seed(111)
pois <- rpois(100, lambda = 3)
binom <- rbinom(100, 20, 0.2)
nbinom <- rnbinom(100, 20, 0.2)
norm <- rnorm(100)

y<-seq(1:100)
plot(y, dbinom(y, 100, 0.1), type = "l", col="salmon", ylim = c(0,.15))
lines(y, dnbinom(y, 100, .9), col="slateblue", ylab="")
lines(y, dnorm(y, mean=50, sd = 4), col="gray", lwd=2)

```


**Main question is** how to identify the number of haplos present that we see across multiple loci.  Parameters that control this are both biological and technological. 

 - how much variation? (*biological*)
 - how many individuals are present in a sample (*biological*)?
 - how much coverage? if 1x coverage, even if you have 1000 haplos may not pick up anything (*technological*)
 
**Coverage and Length Matter Most**

 - coverage and length matter, number of loci doesn't matter as much
 - No of individuals in a sample will effect number of haplos in sample
 - Amount of diversity (relates to read length, more reads means more potential haplos)
 - Coverage will pick up more or less haplos

**Technical vs biological issues** 

 - readout, distribution shifts with variation among indivs, but also with tech issues (coverage, loci length)
 number of haplos vs. number of loci
 - that distrigution is effected by technological and biological
 - variation in number of haplos in a sample is biological

### Review

 - $\mu$ = mutation rate, we assumed general mutation rate of 1e^-8^
 - N~e~ = Coalescent effective size, using conservative est. of 25,000
 - conclusion 3: I think our upward loci length of 10,000 was more of a "future" estimate. We are close with single read options but we prob aren't there yet. 
 - thinking about $\theta$ = 4N$\mu$ can be misleading because N~e~ is coalescent effective population size which is usually orders of magnitude bigger than local census size. For many species our highest theta (10) would be more like a 2 kb DNA fragment. Mean read lengths of 5+ kb are available now with PacBio, but the output needs to increase to make it less cost prohibitive. The limitation will become DNA fragment length in eDNA samples at some point, if not already, even as sequencing tech improves. 
 
 
# Meeting 2021-07-30
 
 - use Tiger Salamanders (see annotated Axolotl genome)
 - using empirical diversity from sequenced individuals in sample population, can get haplotype diversity from that distribution
 - sequence individuals, get pooled haplotype diversity and assess distribution that way, vs. coalescent
 - Ultimately its based on allele freq **distribution** in population sample, which relates to evolutionary scale and is a function of coalescent effective size. Most allele freq distributions will usually have neutral distributions, even if small/divergent populations/isolated. (Mike knows citations to support this)
 - Compare both simulated coalescent with empirical, may be a 10% bias? Don't know.
 - In paper, discuss the potential to use empirical data to inform when possible?

### Sequencing: 

 1. Need to identify target bait/loci in salamander. Based on Melsted/Shaffer paper, only 300-400 loci was enough to estimate k=15 groups with fairly low error rate.
    - Identify sfb1-RAD sites in genome (better for this because they align at same location (no offsets) and they thus have a better pileup and better depth).
    - Search genome for sbf1 sites, pull 10k from either side, then feed that information into REPEATMASKER
    - Reduce masked alignment and align back to genome, make sure only 1x alignments (no dups) and then randomly select 500-1000 loci to use as baits.
    - Use this in low-input prep and then sequence (use MiSeq 600-1000bp reads?)
    
### Lab/Bucket Experiment

 - Depends on what Megan has available, but ideally have something like: 1, 2, 5, 10, 20, sequence 20 individuals (from the 10 bucket, but both replicates).
 - Fine if we take one eDNA filter from each bucket replicate as long as we mix the water prior to filtering. So only "biological" replicates in this case:
 
  - 2 buckets w 1 indiv
  - 2 buckets w 2 indiv
  - 2 buckets w 5 indiv...
  

# Meeting 2022-03-09

Shifting gears to see if we can use eDNA for assessing the number of individuals. Important to do experiments to see what buffers/temperatures/blockers may help optimize eDNA samples for sequencing and capture. 

Really do need a capture step as this was key piece of what we promised, particularly given eDNA sample is <1% of actual target species DNA, and unclear fragmented to what degree.

## Experiment Gear Needed

Assume 3-4 levels depending on fish sampled:

 - **1, 2, 5, 10+**
 - 2 replicates minimum

Gear we'd need to do experiment with stickleback:

 - glass stir rods
 - clean buckets (6 minimum)
 - 5 liters of water
 - clean bubblers, tubing, etc
 - Filters (0.45 micron?)
 - pump and tubing
 - aquarium nets (1 per treatment)
 - water (probably 1 gallon per bucket)?
 - Filter 250ml to 1 L of water per filter (depends on muck)

## Questions? 

 - Wait 2-3 hours per replicate, start with all the fish first then move to the smaller increments?
 - use same fish from large treatment in small treatment? Any issues with this? 
 

Any